### ISSUE TITLE
False warning: Returning Any from function declared to return "bool"

### ISSUE THREAD

**Issue by stefanhoelzl** [python/mypy#5697]
* Are you reporting a bug, or opening a feature request? **Bugreport**
* Please insert below the code you are checking with mypy
```python
from typing import Any

class Structure():
    def __eq__(self, other: Any) -> bool:
        return {} == other
```
* What is the actual behavior/output? 
```console
mypytest.py:5: warning: Returning Any from function declared to return "bool"
```
* What is the behavior/output you expect? **No error**
* What are the versions of mypy and Python you are using?
```console
Python 3.6.5 (v3.6.5:f59c0932b4, Mar 28 2018, 05:52:31) 
[GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.57)] on darwin
```
`mypy==0.630`
* What are the mypy flags you are using?
`--strict`

**Comment by ilevkivskyi** [comment]
Why do you think the warning is false? It looks like a totally reasonable warning, since `--warn-return-any` is part of `--strict`. If you don't like it, don't use this flag.

**Comment by stefanhoelzl** [comment]
because `{} == other` should evaluate to a `bool`.
Which means the method returns a `bool`.
Then getting a warning `Returning Any` seems false to me.

**Comment by ilevkivskyi** [comment]
Everything applied to `Any` evaluates to `Any`. A dynamic class can define `__eq__` that returns a string, there is no guarantee it is a `bool`.

**Comment by ylathouris** [comment]
> Everything applied to `Any` evaluates to `Any`. A dynamic class can define `__eq__` that returns a string, there is no guarantee it is a `bool`.

This was helpful. Thank you!

**Comment by Torxed** [comment]
I get the same error when using:
```python
def verify_password(plain_password :str, hashed_password :str) -> bool:
	return pwd_context.verify(plain_password, hashed_password)
```
Which is taken from the [FastAPI](https://fastapi.tiangolo.com/tutorial/security/oauth2-jwt/) docs. In this case `.verify()` is a verification function that to the best of my knowledge can only return `True` or `False` but is not typed: https://github.com/efficks/passlib/blob/bf46115a2d4d40ec7901eb6982198fd82cc1d6f4/passlib/context.py#L1832-L1912

I think I understand the rules of the `Any` evaluation, but perhaps I've misunderstood how mypy evaluates these cases.
Shouldn't the return value matter in mypy's checks? That is that it's a static value of `False` or `True` that gets returned?

In any case, the fix is simple enough:
```python
def verify_password(plain_password :str, hashed_password :str) -> bool:
	return bool(pwd_context.verify(plain_password, hashed_password))
```
But it feels excessive to enforce one additional function call per call stack to please mypy, even in `--strict` mode.
It's the first time trying to use linters and type checkers, and this thread came up. I assumed it would deal with these cases "smartly", is that not the case?

**Comment by JelleZijlstra** [comment]
@Torxed the fix seems simple here: mypy needs to be told what the return type of `.verify()` is. If there's no annotation, it doesn't know.

I would not recommend turning on the option that generates this error. It's too strict to be useful for most code.

**Comment by hauntsaninja** [comment]
`--no-warn-return-any` :-)

**Comment by mattmontero** [comment]
same issue here 👋 

My little hack for this is to set a new variable, and type the variable. (Similar to @Torxed, but not explicitly casting)

``` python
from typing import Any

class Structure():
    def __eq__(self, other: Any) -> bool:
        outcome: bool = {} == other
        return outcome
```

**Comment by Torxed** [comment]
> > @Torxed the fix seems simple here: mypy needs to be told what the return type of `.verify()` is. If there's no annotation, it doesn't know.
> 
> The same problem affects me. It is evident that `.verify()` returns a boolean. Why is it invisible to Mypy?
> 
> ![image](https://private-user-images.githubusercontent.com/85762931/334199881-65118a05-194e-439c-ab37-2ba747dac292.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTY4NDM0OTEsIm5iZiI6MTcxNjg0MzE5MSwicGF0aCI6Ii84NTc2MjkzMS8zMzQxOTk4ODEtNjUxMThhMDUtMTk0ZS00MzljLWFiMzctMmJhNzQ3ZGFjMjkyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTI3VDIwNTMxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ3NDc5ZGQ5NWNiY2M5ZTZkNDZkNzJjZDVhN2E2NDA3NDg0NDRkZDQ1ZTFmZTQ2MjUwNDVjMjYzOGE5Y2ZkMDkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.kqANDmlKWT0AP7LxlvXuvLlqeLaF3Wcbq8IYrX6eZ2E)

It's actually not completely evident. Because the [`.verify()`](https://foss.heptapod.net/python-libs/passlib/-/blob/01dfe753e4f4590307611ef5c6304279a7364305/passlib/context.py?page=3#L2272-2347) function actually lacks a typed annotation that mypy can interpret.
Your IDE however, might be able to parse the [sphinx docstring](https://sphinx-rtd-tutorial.readthedocs.io/en/latest/docstrings.html#the-sphinx-docstring-format) since the verify docstring has a `:returns:` or, if exist, `:rtype:`.

But to the best of my understanding, passlib simply lacks native annotation in [all possible `.verify()` functions](https://foss.heptapod.net/search?search=def%20verify&nav_source=navbar&project_id=142&group_id=379&search_code=true&repository_ref=01dfe753e4f4590307611ef5c6304279a7364305).

And to the best of my understanding, `mypy` does not read or understand docstrings, and why would it, as it's a [sphinx invented thing](https://sphinx-rtd-tutorial.readthedocs.io/en/latest/docstrings.html#the-sphinx-docstring-format) *(as I understand it)*.

Would be nice if the project moved to https://pypi.org/project/sphinx-autodoc-typehints/ or something similar.
